### PR TITLE
PHP Integration

### PR DESCRIPTION
**Pull request by tuxology**
* integrate PHP Frontend from upstream Joern into Privado
* vendor PHP Parser from upstream fork into Privado and use it 
* Use 4GB heap size for the tests to run in CI.

### CODE REVIEW

**Review by khemrajrathore — COMMENTED** [review]
Overall changes look good to me, If we can also have some test cases around PHP parsing and tagging that would be great

**Comment by karan-batavia** [comment]
Can we please have some tests on the basic tagging process? Let's ensure the integration stays solid from the start itself.

**Comment by pawan-privado** [comment]
> Can we please have some tests on the basic tagging process? Let's ensure the integration stays solid from the start itself. @pawan-privado

Sure!

**Comment by karan-batavia** [comment]
Also, When are we planning to release these changes? Let's make sure we don't conflict with the C# release on Monday, 
@pawan-privado @tuxology

File: build.sbt [hunk 1]
```diff
@@ -147,7 +148,7 @@ goAstGenDlTask := {
   distDir.listFiles().foreach(_.setExecutable(true, false))
 }
 
-Compile / compile := ((Compile / compile) dependsOn goAstGenDlTask).value
+Compile / compile := (Compile / compile).dependsOn(goAstGenDlTask, phpParseDlTask).value
```
Comment (karan-batavia) [inline build.sbt:151]: This doesn't work on Linux, can you please test once and confirm? @pawan-privado

Comment (pawan-privado) [reply build.sbt:151]: I don't have access to a Linux machine ATM but it works fine on macOS. I've DM'd you something on Slack, btw, regarding this.

**Comment by pawan-privado** [comment]
@tuxology Test failed because PHP parser could not be executed.
```
Error:  Invalid path for PhpParserBin: /home/runner/.cache/coursier/v1/https/repo1.maven.org/maven2/io/joern/php2cpg/bin/php-parser/php-parser.php
```

The path doesn't look right to me. Is this the intended path for vendoring the parser?

File: build.sbt (hunk 1 above)
Comment (tuxology) [reply build.sbt]: This looked fine to me - I tested on Linux, but did a cleaner version of this just in case

**Comment by tuxology** [comment]
> @tuxology Test failed because PHP parser could not be executed.
> 
> ```
> Error:  Invalid path for PhpParserBin: /home/runner/.cache/coursier/v1/https/repo1.maven.org/maven2/io/joern/php2cpg/bin/php-parser/php-parser.php
> ```
> 
> The path doesn't look right to me. Is this the intended path for vendoring the parser?

@pawan-privado It looks like this is only happening in Github - the path for `php-parser` is possibly null and hence the default path is being picked (which is not vendored path). This was working before C# push (https://github.com/Privado-Inc/privado-core/pull/983/commits/a883357d7bdf73735da07bec602850f1454446ae) so if you can investigate and fix, it would help 👍

File: src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        new File(dir.substring("file:".length, indexOfLib)).toString
+      } else {
+        val indexOfTarget = dir.lastIndexOf("target")
+        if (indexOfTarget != -1) {
+          new File(dir.substring("file:".length, indexOfTarget)).toString
+        } else {
+          "."
+        }
+      }
+      Paths.get(fixedDir, "/bin/php-parser/php-parser.php").toAbsolutePath.toString
```
Comment (pawan-privado) [inline src/main/scala/ai/privado/languageEngine/php/processor/PhpProcessor.scala:87]: 1. This path looks suspiciously wrong. I have a `bin` folder in my parser which has a single file called `php-parse`.
2. You need to run `composer install` prior to using the parser.

**Comment by pawan-privado** [comment]
> > @tuxology Test failed because PHP parser could not be executed.
> > ```
> > Error:  Invalid path for PhpParserBin: /home/runner/.cache/coursier/v1/https/repo1.maven.org/maven2/io/joern/php2cpg/bin/php-parser/php-parser.php
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The path doesn't look right to me. Is this the intended path for vendoring the parser?
> 
> @pawan-privado It looks like this is only happening in Github - the path for `php-parser` is possibly null and hence the default path is being picked (which is not vendored path). This was working before C# push ([a883357](https://github.com/Privado-Inc/privado-core/commit/a883357d7bdf73735da07bec602850f1454446ae)) so if you can investigate and fix, it would help 👍

The reason why we did not notice this happen in our dev machines is because we had the required `env` variable set and the path was being picked up from the variable.

File: build.sbt (hunk 1 above)
Comment (pawan-privado) [reply build.sbt]: > This doesn't work on Linux, can you please test once and confirm? @pawan-privado

Haven't noticed any abnormalities at my end. Can you attach the error log if it still continues to happen? AFAIK, there's no platform specific stage in our build process so it shouldn't fail on Linux systems.

File: src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          |
+          |  function add_phone($ph) {
+          |    //
+          |  }
+          |}
+          |?>
+          |""".stripMargin)
+
+      val List(firstName) = cpg.member.l
+      firstName.tag.nameExact(Constants.catLevelOne).value.l shouldBe List(CatLevelOne.SOURCES.name)
```
Comment (karan-batavia) [inline src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala:49]: Are we testing field identifiers here, or members? I don't see a field access node being accessed.

Comment (karan-batavia) [reply src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala:49]: You can refer C# Field Identifier Tests

Comment (pawan-privado) [reply src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala]: Fixed.

File: src/test/scala/ai/privado/languageEngine/php/tagger/source/IdentifierTaggingTest.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        |  function get_email() {
+        |    return $emailId;
+        |  }
+        |}
+        |?>
+        |
+        |""".stripMargin)
+
+    "tag member in a structure" in {
+      cpg.member("firstName").tag.nameExact(Constants.id).value.l shouldBe List(
```
Comment (karan-batavia) [inline src/test/scala/ai/privado/languageEngine/php/tagger/source/IdentifierTaggingTest.scala:53]: Let's test for identifiers as well here.

File: src/test/scala/ai/privado/languageEngine/php/tagger/source/IdentifierTaggingTest.scala [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        "Data.Sensitive.PersonalIdentification.FirstName"
+      )
+
+      cpg.member("dob").tag.nameExact(Constants.id).value.l shouldBe List(
```
Comment (karan-batavia) [inline src/test/scala/ai/privado/languageEngine/php/tagger/source/IdentifierTaggingTest.scala:57]: Here as well.

**Review by kbatavia12 — APPROVED** [review]
LGTM

**Review by karan-batavia — DISMISSED** [review]

**Review by karan-batavia — DISMISSED** [review]

File: src/test/scala/ai/privado/languageEngine/php/PhpTestBase.scala [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+import org.scalatest.{BeforeAndAfterAll, BeforeAndAfterEach}
+
+import scala.collection.mutable
+
+abstract class PhpTestBase extends AnyWordSpec with Matchers with BeforeAndAfterAll with BeforeAndAfterEach {
+  private val cpgs        = mutable.ArrayBuffer.empty[Cpg]
+  private val outPutFiles = mutable.ArrayBuffer.empty[File]
+  private val inputDirs   = mutable.ArrayBuffer.empty[File]
+
+  private val sourceRules = List(
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/languageEngine/php/PhpTestBase.scala:48]: The test source rules are defined at https://github.com/Privado-Inc/privado-core/blob/main/src/test/scala/ai/privado/RuleInfoTestData.scala kindly use that

File: src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+          |
+          |  function add_phone($ph) {
+          |    //
+          |  }
+          |}
+          |?>
+          |""".stripMargin)
+
+      val List(firstNameField) = cpg.fieldAccess.l
+      firstNameField.tag.nameExact(Constants.catLevelOne).value.l shouldBe List(CatLevelOne.SOURCES.name)
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/languageEngine/php/tagger/source/FieldIdentifierTaggingTests.scala:50]: also add a check on the `code` of `firstNameField` to assert we are getting the expected fieldAccess tagged

File: src/test/scala/ai/privado/languageEngine/php/tagger/source/LiteralTaggingTests.scala [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+          |  }
+          |
+          |  function add_phone($ph) {
+          |  }
+          |}
+          |?>
+          |""".stripMargin)
+
+      val literals = cpg.literal.l
+      literals.last.tag.nameExact(Constants.catLevelOne).value.l shouldBe List(CatLevelOne.SOURCES.name)
```
Comment (khemrajrathore) [inline src/test/scala/ai/privado/languageEngine/php/tagger/source/LiteralTaggingTests.scala:46]: Also add a check on the `code` of literal to assert we are tagging the right literal

**Comment by khemrajrathore** [comment]
@pawan-privado Take a merge of dev, they are are good to go

**Review by khemrajrathore — APPROVED** [review]
LGTM

**Pull request merged**